From fb9baf2f0f40c2517ad3b0ec9955b0ef51af26e9 Mon Sep 17 00:00:00 2001
From: Cong Nguyen <cong.nguyen.zn@renesas.com>
Date: Tue, 15 Apr 2025 03:25:40 +0000
Subject: [PATCH 10/16] Add support for CRU driver

Currently this driver is disable in the dts file.
Set status to "okay" to enable CRU driver.

Upstream-Status: Pending
---
 arch/arm64/boot/dts/renesas/r9a09g057.dtsi    | 228 +++++++++
 .../boot/dts/renesas/r9a09g057h4-evk-ver1.dts | 247 ++++++++++
 arch/arm64/configs/defconfig                  |   1 +
 drivers/gpio/Kconfig                          |   6 +
 .../platform/renesas/rzg2l-cru/rzg2l-core.c   | 143 +++++-
 .../platform/renesas/rzg2l-cru/rzg2l-cru.h    |  71 ++-
 .../platform/renesas/rzg2l-cru/rzg2l-csi2.c   | 117 ++++-
 .../platform/renesas/rzg2l-cru/rzg2l-ip.c     |  13 +-
 .../platform/renesas/rzg2l-cru/rzg2l-video.c  | 455 +++++++++++++-----
 drivers/pinctrl/renesas/Kconfig               |   1 +
 10 files changed, 1142 insertions(+), 140 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r9a09g057.dtsi b/arch/arm64/boot/dts/renesas/r9a09g057.dtsi
index 18293fb55..e9b5177a0 100644
--- a/arch/arm64/boot/dts/renesas/r9a09g057.dtsi
+++ b/arch/arm64/boot/dts/renesas/r9a09g057.dtsi
@@ -869,6 +869,234 @@ eth1: ethernet@15C40000 {
 			snps,burst-map = <0x7>;
 		};
 
+		cru0: video0@16000000 {
+ 			compatible = "renesas,cru-r9a09g057", "renesas,rzv2h-cru";
+ 			reg = <0 0x16000000 0 0x400>;
+ 			clocks = <&cpg CPG_MOD 211>,
+ 				 <&cpg CPG_MOD 212>,
+ 				 <&cpg CPG_MOD 210>;
+ 			clock-names = "video", "apb", "axi";
+ 			interrupts = <GIC_SPI 838 IRQ_TYPE_LEVEL_HIGH>;
+ 			resets = <&cpg 197>,
+ 				 <&cpg 198>;
+ 			reset-names = "presetn", "aresetn";
+ 			power-domains = <&cpg>;
+ 			channel,id = <0>;
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					cru0_to_csi20: endpoint {
+ 						remote-endpoint =
+ 							<&csi20_to_cru0>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		cru1: video1@16010000 {
+ 			compatible = "renesas,cru-r9a09g057", "renesas,rzv2h-cru";
+ 			reg = <0 0x16010000 0 0x400>;
+ 			clocks = <&cpg CPG_MOD 214>,
+ 				 <&cpg CPG_MOD 215>,
+ 				 <&cpg CPG_MOD 213>;
+ 			clock-names = "video", "apb", "axi";
+ 			interrupts = <GIC_SPI 844 IRQ_TYPE_LEVEL_HIGH>;
+ 			resets = <&cpg 200>,
+ 				 <&cpg 201>;
+ 			reset-names = "presetn", "aresetn";
+ 			power-domains = <&cpg>;
+ 			channel,id = <1>;
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					cru1_to_csi21: endpoint {
+ 						remote-endpoint =
+ 							<&csi21_to_cru1>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		cru2: video2@16020000 {
+ 			compatible = "renesas,cru-r9a09g057", "renesas,rzv2h-cru";
+ 			reg = <0 0x16020000 0 0x400>;
+ 			clocks = <&cpg CPG_MOD 217>,
+ 				 <&cpg CPG_MOD 218>,
+ 				 <&cpg CPG_MOD 216>;
+ 			clock-names = "video", "apb", "axi";
+ 			interrupts = <GIC_SPI 850 IRQ_TYPE_LEVEL_HIGH>;
+ 			resets = <&cpg 203>,
+ 				 <&cpg 204>;
+ 			reset-names = "presetn", "aresetn";
+ 			power-domains = <&cpg>;
+ 			channel,id = <2>;
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					cru2_to_csi22: endpoint {
+ 						remote-endpoint =
+ 							<&csi22_to_cru2>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		cru3: video3@16030000 {
+ 			compatible = "renesas,cru-r9a09g057", "renesas,rzv2h-cru";
+ 			reg = <0 0x16030000 0 0x400>;
+ 			clocks = <&cpg CPG_MOD 220>,
+ 				 <&cpg CPG_MOD 221>,
+ 				 <&cpg CPG_MOD 219>;
+ 			clock-names = "video", "apb", "axi";
+ 			interrupts = <GIC_SPI 855 IRQ_TYPE_LEVEL_HIGH>;
+ 			resets = <&cpg 206>,
+ 				 <&cpg 207>;
+ 			reset-names = "presetn", "aresetn";
+ 			power-domains = <&cpg>;
+ 			channel,id = <3>;
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					cru3_to_csi23: endpoint {
+ 						remote-endpoint =
+ 							<&csi23_to_cru3>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		csi20: csi20@16000400 {
+ 			compatible = "renesas,r9a09g057-csi2", "renesas,rzv2h-csi2";
+ 			reg = <0 0x16000400 0 0xc00>;
+ 			interrupts = <GIC_SPI 837 IRQ_TYPE_LEVEL_HIGH>;
+ 			clocks = <&cpg CPG_MOD 211>,
+ 				 <&cpg CPG_MOD 212>;
+ 			clock-names = "video", "apb";
+ 			resets = <&cpg 197>,
+ 				 <&cpg 199>;
+ 			reset-names ="presetn", "cmn-rstb";
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					csi20_to_cru0: endpoint {
+ 						remote-endpoint =
+ 							<&cru0_to_csi20>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		csi21: csi21@16010400 {
+ 			compatible = "renesas,r9a09g057-csi2", "renesas,rzv2h-csi2";
+ 			reg = <0 0x16010400 0 0xc00>;
+ 			interrupts = <GIC_SPI 843 IRQ_TYPE_LEVEL_HIGH>;
+ 			clocks = <&cpg CPG_MOD 214>,
+ 				 <&cpg CPG_MOD 215>;
+ 			clock-names = "video", "apb";
+ 			resets = <&cpg 200>,
+ 				 <&cpg 202>;
+ 			reset-names = "presetn", "cmn-rstb";
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					csi21_to_cru1: endpoint {
+ 						remote-endpoint =
+ 							<&cru1_to_csi21>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		csi22: csi22@16020400 {
+ 			compatible = "renesas,r9a09g057-csi2", "renesas,rzv2h-csi2";
+ 			reg = <0 0x16020400 0 0xc00>;
+ 			interrupts = <GIC_SPI 849 IRQ_TYPE_LEVEL_HIGH>;
+ 			clocks = <&cpg CPG_MOD 217>,
+ 				 <&cpg CPG_MOD 218>;
+ 			clock-names = "video", "apb";
+ 			resets = <&cpg 203>,
+ 				 <&cpg 205>;
+ 			reset-names = "presetn", "cmn-rstb";
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					csi22_to_cru2: endpoint {
+ 						remote-endpoint =
+ 							<&cru2_to_csi22>;
+ 					};
+ 				};
+ 			};
+ 		};
+ 
+ 		csi23: csi23@16030400 {
+ 			compatible = "renesas,r9a09g057-csi2", "renesas,rzv2h-csi2";
+ 			reg = <0 0x16030400 0 0xc00>;
+ 			interrupts = <GIC_SPI 854 IRQ_TYPE_LEVEL_HIGH>;
+ 			clocks = <&cpg CPG_MOD 220>,
+ 				 <&cpg CPG_MOD 221>;
+ 			clock-names = "video", "apb";
+ 			resets = <&cpg 206>,
+ 				 <&cpg 208>;
+ 			reset-names ="presetn", "cmn-rstb";
+ 			status = "disabled";
+ 
+ 			ports {
+ 				#address-cells = <1>;
+ 				#size-cells = <0>;
+ 
+ 				port@1 {
+ 					reg = <1>;
+ 
+ 					csi23_to_cru3: endpoint {
+ 						remote-endpoint =
+ 							<&cru3_to_csi23>;
+ 					};
+ 				};
+ 			};
+ 		};
+
 		ostm0: timer@11800000 {
 			compatible = "renesas,r9a09g057-ostm", "renesas,ostm";
 			reg = <0x0 0x11800000 0x0 0x1000>;
diff --git a/arch/arm64/boot/dts/renesas/r9a09g057h4-evk-ver1.dts b/arch/arm64/boot/dts/renesas/r9a09g057h4-evk-ver1.dts
index d558931ab..bac4f8060 100644
--- a/arch/arm64/boot/dts/renesas/r9a09g057h4-evk-ver1.dts
+++ b/arch/arm64/boot/dts/renesas/r9a09g057h4-evk-ver1.dts
@@ -132,6 +132,43 @@ vqmmc_sdhi1: regulator-vccq-sdhi1 {
 		states = <3300000 0 1800000 1>;
 	};
 
+	ov5645_fixed_clk: osc25250-clk {
+ 		compatible = "fixed-clock";
+ 		#clock-cells = <0>;
+ 		clock-frequency = <24000000>;
+ 	};
+ 
+ 	ov5645_vdddo_1v8: 1p8v {
+ 		compatible = "regulator-fixed";
+ 		regulator-name = "camera_vdddo";
+ 		regulator-min-microvolt = <1800000>;
+ 		regulator-max-microvolt = <1800000>;
+ 		regulator-always-on;
+ 	};
+ 
+ 	ov5645_vdda_2v8: 2p8v {
+ 		compatible = "regulator-fixed";
+ 		regulator-name = "camera_vdda";
+ 		regulator-min-microvolt = <2800000>;
+ 		regulator-max-microvolt = <2800000>;
+ 		regulator-always-on;
+ 	};
+ 
+ 	ov5645_vddd_1v5: 1p5v {
+ 		compatible = "regulator-fixed";
+ 		regulator-name = "camera_vddd";
+ 		regulator-min-microvolt = <1500000>;
+ 		regulator-max-microvolt = <1500000>;
+ 		regulator-always-on;
+ 	};
+ 
+ 	gpio_dummy: gpio-dummy {
+ 		compatible = "gpio-dummy";
+ 		gpio-controller;
+ 		#gpio-cells = <2>;
+ 		ngpios = <2>;
+ 	};
+
 	osc1: cec-clock {
 		compatible = "fixed-clock";
 		#clock-cells = <0>;
@@ -169,6 +206,14 @@ sd1_pwr_en {
 		line-name = "sd1_pwr_en";
 	};
 
+	/* cam0_pwr_en is used by Google Coral Camera from CSI20 and CSI21 */
+ 	cam0_pwr_en {
+ 		gpio-hog;
+ 		gpios = <RZV2H_GPIO(B, 2) GPIO_ACTIVE_HIGH>;
+ 		output-high;
+ 		line-name = "cam0_pwr_en";
+ 	};
+
 	sdhi0_pins: sd0 {
 		sd0_mux {
 			pinmux = <RZV2H_PORT_PINMUX(A, 5, 15)>; /* SD0_CD */
@@ -201,6 +246,26 @@ sd1_cd {
 		};
 	};
 
+	i2c0_pins: i2c0 {
+ 		pinmux = <RZV2H_PORT_PINMUX(3, 0, 1)>, /* I2C0_SDA */
+ 			 <RZV2H_PORT_PINMUX(3, 1, 1)>; /* I2C0_SCL */
+ 	};
+ 
+ 	i2c1_pins: i2c1 {
+ 		pinmux = <RZV2H_PORT_PINMUX(3, 2, 1)>, /* I2C1_SDA */
+ 			 <RZV2H_PORT_PINMUX(3, 3, 1)>; /* I2C1_SCL */
+ 	};
+ 
+ 	i2c6_pins: i2c6 {
+ 		pinmux = <RZV2H_PORT_PINMUX(4, 4, 1)>, /* I2C6_SDA */
+ 			 <RZV2H_PORT_PINMUX(4, 5, 1)>; /* I2C6_SCL */
+ 	};
+ 
+ 	i2c7_pins: i2c7 {
+ 		pinmux = <RZV2H_PORT_PINMUX(4, 6, 1)>, /* I2C7_SDA */
+ 			 <RZV2H_PORT_PINMUX(4, 7, 1)>; /* I2C7_SCL */
+ 	};
+
 	usb30_pins: usb30 {
 		pinmux = <RZV2H_PORT_PINMUX(B, 0, 14)>; /* VBUS */
 	};
@@ -324,6 +389,188 @@ phy1: ethernet-phy@0 {
 	};
 };
 
+&i2c0 {
+ 	pinctrl-0 = <&i2c0_pins>;
+ 	pinctrl-names = "default";
+ 
+ 	status = "disabled";
+ 
+ 	ov5645_csi20: ov5645_csi20@3c {
+ 		compatible = "ovti,ov5645";
+ 		reg = <0x3c>;
+ 		clocks = <&ov5645_fixed_clk>;
+ 		clock-frequency = <24000000>;
+ 		vdddo-supply = <&ov5645_vdddo_1v8>;
+ 		vdda-supply = <&ov5645_vdda_2v8>;
+ 		vddd-supply = <&ov5645_vddd_1v5>;
+ 		/* Set dummy enable gpio */
+ 		enable-gpios = <&gpio_dummy 0 GPIO_ACTIVE_HIGH>;
+ 		reset-gpios = <&pinctrl RZV2H_GPIO(6, 4) GPIO_ACTIVE_LOW>;
+ 
+ 		port {
+ 			ov5645_to_csi20: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&csi20_in>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&i2c1 {
+ 	pinctrl-0 = <&i2c1_pins>;
+ 	pinctrl-names = "default";
+ 
+ 	status = "disabled";
+ 
+ 	ov5645_csi21: ov5645_csi21@3c {
+ 		compatible = "ovti,ov5645";
+ 		reg = <0x3c>;
+ 		clocks = <&ov5645_fixed_clk>;
+ 		clock-frequency = <24000000>;
+ 		vdddo-supply = <&ov5645_vdddo_1v8>;
+ 		vdda-supply = <&ov5645_vdda_2v8>;
+ 		vddd-supply = <&ov5645_vddd_1v5>;
+ 		/* Set dummy enable gpio */
+ 		enable-gpios = <&gpio_dummy 1 GPIO_ACTIVE_HIGH>;
+ 		reset-gpios = <&pinctrl RZV2H_GPIO(6, 5) GPIO_ACTIVE_LOW>;
+ 
+ 		port {
+ 			ov5645_to_csi21: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&csi21_in>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&i2c6 {
+ 	pinctrl-0 = <&i2c6_pins>;
+ 	pinctrl-names = "default";
+ 
+ 	status = "disabled";
+ 
+ 	ov5645_csi22: ov5645_csi22@3c {
+ 		compatible = "ovti,ov5645";
+ 		reg = <0x3c>;
+ 		clocks = <&ov5645_fixed_clk>;
+ 		clock-frequency = <24000000>;
+ 		vdddo-supply = <&ov5645_vdddo_1v8>;
+ 		vdda-supply = <&ov5645_vdda_2v8>;
+ 		vddd-supply = <&ov5645_vddd_1v5>;
+ 		enable-gpios = <&pinctrl RZV2H_GPIO(6, 1) GPIO_ACTIVE_HIGH>;
+ 		reset-gpios = <&pinctrl RZV2H_GPIO(9, 2) GPIO_ACTIVE_LOW>;
+ 
+ 		port {
+ 			ov5645_to_csi22: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&csi22_in>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&i2c7 {
+ 	pinctrl-0 = <&i2c7_pins>;
+ 	pinctrl-names = "default";
+ 
+ 	status = "disabled";
+ 
+ 	ov5645_csi23: ov5645_csi23@3c {
+ 		compatible = "ovti,ov5645";
+ 		reg = <0x3c>;
+ 		clocks = <&ov5645_fixed_clk>;
+ 		clock-frequency = <24000000>;
+ 		vdddo-supply = <&ov5645_vdddo_1v8>;
+ 		vdda-supply = <&ov5645_vdda_2v8>;
+ 		vddd-supply = <&ov5645_vddd_1v5>;
+ 		enable-gpios = <&pinctrl RZV2H_GPIO(6, 0) GPIO_ACTIVE_HIGH>;
+ 		reset-gpios = <&pinctrl RZV2H_GPIO(9, 3) GPIO_ACTIVE_LOW>;
+ 
+ 		port {
+ 			ov5645_to_csi23: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&csi23_in>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&cru0 {
+ 	status = "okay";
+};
+ 
+&csi20 {
+ 	status = "disabled";
+ 
+ 	ports {
+ 		port {
+ 			csi20_in: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&ov5645_to_csi20>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&cru1 {
+ 	status = "okay";
+};
+ 
+&csi21 {
+ 	status = "disabled";
+ 
+ 	ports {
+ 		port {
+ 			csi21_in: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&ov5645_to_csi21>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&cru2 {
+ 	status = "okay";
+};
+ 
+&csi22 {
+ 	status = "disabled";
+ 
+ 	ports {
+ 		port {
+ 			csi22_in: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&ov5645_to_csi22>;
+ 			};
+ 		};
+ 	};
+};
+ 
+&cru3 {
+ 	status = "okay";
+};
+ 
+&csi23 {
+ 	status = "disabled";
+ 
+ 	ports {
+ 		port {
+ 			csi23_in: endpoint {
+ 				clock-lanes = <0>;
+ 				data-lanes = <1 2>;
+ 				remote-endpoint = <&ov5645_to_csi23>;
+ 			};
+ 		};
+ 	};
+};
+
 &ostm0 {
 	status = "okay";
 };
diff --git a/arch/arm64/configs/defconfig b/arch/arm64/configs/defconfig
index 823eb5c01..f348107f8 100644
--- a/arch/arm64/configs/defconfig
+++ b/arch/arm64/configs/defconfig
@@ -171,6 +171,7 @@ CONFIG_SPI_RPCIF=y
 CONFIG_PINCTRL_SINGLE=y
 CONFIG_PINCTRL_DA9062=y
 CONFIG_GPIO_RCAR=y
+CONFIG_GPIO_DUMMY=y
 CONFIG_GPIO_PCA953X=y
 CONFIG_GPIO_PCA953X_IRQ=y
 CONFIG_GPIO_SYSFS=y
diff --git a/drivers/gpio/Kconfig b/drivers/gpio/Kconfig
index 1c28a4891..74223761e 100644
--- a/drivers/gpio/Kconfig
+++ b/drivers/gpio/Kconfig
@@ -562,6 +562,12 @@ config GPIO_RCAR
 	help
 	  Say yes here to support GPIO on Renesas R-Car or RZ/G SoCs.
 
+config GPIO_DUMMY
+ 	tristate "Renesas Dummy GPIO driver"
+ 	depends on ARCH_RENESAS || COMPILE_TEST
+ 	help
+ 	  Say yes here to build support for the Dummy GPIO.
+
 config GPIO_RDA
 	bool "RDA Micro GPIO controller support"
 	depends on ARCH_RDA || COMPILE_TEST
diff --git a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-core.c b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-core.c
index 4d499dae1..97bc5d123 100644
--- a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-core.c
+++ b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-core.c
@@ -73,6 +73,7 @@ static int rzg2l_cru_group_notify_complete(struct v4l2_async_notifier *notifier)
 		return ret;
 	}
 	cru->csi.channel = 0;
+	cru->svc_channel = cru->csi.channel;
 	cru->ip.remote = cru->csi.subdev;
 
 	/* Create media device link between CRU IP <-> CRU OUTPUT */
@@ -242,6 +243,24 @@ static int rzg2l_cru_media_init(struct rzg2l_cru_dev *cru)
 	if (ret)
 		return ret;
 
+	if (cru->info->max_cru_channels > 1) {
+		ret = of_property_read_u32(cru->dev->of_node, "channel,id", &cru->id);
+		if (ret) {
+			if (cru->info->cru_type == RZV2H_CRU_TYPE) {
+				dev_err(cru->dev, "%pOF: No channel,id property found\n",
+						cru->dev->of_node);
+				return -EINVAL;
+			}
+			cru->id = 0;
+		}
+	}
+
+	if (cru->id >= cru->info->max_cru_channels) {
+		dev_err(cru->dev, "%pOF: Invalid channel,id '%u'\n",
+			cru->dev->of_node, cru->id);
+		return -EINVAL;
+	}
+
 	mutex_init(&cru->mdev_lock);
 	mdev = &cru->mdev;
 	mdev->dev = cru->dev;
@@ -270,10 +289,11 @@ static int rzg2l_cru_media_init(struct rzg2l_cru_dev *cru)
 static int rzg2l_cru_probe(struct platform_device *pdev)
 {
 	struct rzg2l_cru_dev *cru;
+	struct device *dev = &pdev->dev;
 	struct v4l2_ctrl *ctrl;
 	int ret;
 
-	cru = devm_kzalloc(&pdev->dev, sizeof(*cru), GFP_KERNEL);
+	cru = devm_kzalloc(dev, sizeof(*cru), GFP_KERNEL);
 	if (!cru)
 		return -ENOMEM;
 
@@ -281,23 +301,23 @@ static int rzg2l_cru_probe(struct platform_device *pdev)
 	if (IS_ERR(cru->base))
 		return PTR_ERR(cru->base);
 
-	cru->presetn = devm_reset_control_get_shared(&pdev->dev, "presetn");
+	cru->presetn = devm_reset_control_get_shared(dev, "presetn");
 	if (IS_ERR(cru->presetn))
 		return dev_err_probe(&pdev->dev, PTR_ERR(cru->presetn),
 				     "Failed to get cpg presetn\n");
 
-	cru->aresetn = devm_reset_control_get_exclusive(&pdev->dev, "aresetn");
+	cru->aresetn = devm_reset_control_get_exclusive(dev, "aresetn");
 	if (IS_ERR(cru->aresetn))
 		return dev_err_probe(&pdev->dev, PTR_ERR(cru->aresetn),
 				     "Failed to get cpg aresetn\n");
 
-	cru->vclk = devm_clk_get(&pdev->dev, "video");
+	cru->vclk = devm_clk_get(dev, "video");
 	if (IS_ERR(cru->vclk))
-		return dev_err_probe(&pdev->dev, PTR_ERR(cru->vclk),
+		return dev_err_probe(dev, PTR_ERR(cru->vclk),
 				     "Failed to get video clock\n");
 
-	cru->dev = &pdev->dev;
-	cru->info = of_device_get_match_data(&pdev->dev);
+	cru->dev = dev;
+	cru->info = of_device_get_match_data(dev);
 
 	cru->image_conv_irq = platform_get_irq(pdev, 0);
 	if (cru->image_conv_irq < 0)
@@ -310,8 +330,8 @@ static int rzg2l_cru_probe(struct platform_device *pdev)
 		return ret;
 
 	cru->num_buf = RZG2L_CRU_HW_BUFFER_DEFAULT;
-	pm_suspend_ignore_children(&pdev->dev, true);
-	pm_runtime_enable(&pdev->dev);
+	pm_suspend_ignore_children(dev, true);
+ 	pm_runtime_enable(dev);
 
 	ret = rzg2l_cru_media_init(cru);
 	if (ret)
@@ -344,7 +364,7 @@ static int rzg2l_cru_probe(struct platform_device *pdev)
 	
 error_dma_unregister:
 	rzg2l_cru_dma_unregister(cru);
-	pm_runtime_disable(&pdev->dev);
+	pm_runtime_disable(dev);
 
 	return ret;
 }
@@ -365,8 +385,109 @@ static void rzg2l_cru_remove(struct platform_device *pdev)
 	rzg2l_cru_dma_unregister(cru);
 }
 
+static const u16 rzg2l_cru_regs[CRU_REGS_END] = {
+	[CRUnCTRL] = 0x0,
+	[CRUnIE] = 0x4,
+	[CRUnINTS] = 0x8,
+	[CRUnRST] = 0xC,
+	[AMnMB1ADDRL] = 0x100,
+	[AMnMB1ADDRH] = 0x104,
+	[AMnMB2ADDRL] = 0x108,
+	[AMnMB2ADDRH] = 0x10C,
+	[AMnMB3ADDRL] = 0x110,
+	[AMnMB3ADDRH] = 0x114,
+	[AMnMB4ADDRL] = 0x118,
+	[AMnMB4ADDRH] = 0x11C,
+	[AMnMB5ADDRL] = 0x120,
+	[AMnMB5ADDRH] = 0x124,
+	[AMnMB6ADDRL] = 0x128,
+	[AMnMB6ADDRH] = 0x12C,
+	[AMnMB7ADDRL] = 0x130,
+	[AMnMB7ADDRH] = 0x134,
+	[AMnMB8ADDRL] = 0x138,
+	[AMnMB8ADDRH] = 0x13C,
+	[AMnMBVALID] = 0x148,
+	[AMnMBS] = 0x14C,
+	[AMnAXIATTR] = 0x158,
+	[AMnFIFO] = 0x160,
+	[AMnFIFOPNTR] = 0x168,
+	[AMnAXISTP] = 0x174,
+	[AMnAXISTPACK] = 0x178,
+	[ICnEN] = 0x200,
+	[ICnMC] = 0x208,
+	[ICnMS] = 0x254,
+	[ICnDMR] = 0x26C,
+};
+
+static const u16 rzv2h_cru_regs[CRU_REGS_END] = {
+	[CRUnCTRL] = 0x0,
+	[CRUnIE] = 0x4,
+	[CRUnIE2] = 0x8,
+	[CRUnINTS] = 0xC,
+	[CRUnINTS2] = 0x10,
+	[CRUnRST] = 0x18,
+	[AMnMB1ADDRL] = 0x40,
+	[AMnMB1ADDRH] = 0x44,
+	[AMnMB2ADDRL] = 0x48,
+	[AMnMB2ADDRH] = 0x4C,
+	[AMnMB3ADDRL] = 0x50,
+	[AMnMB3ADDRH] = 0x54,
+	[AMnMB4ADDRL] = 0x58,
+	[AMnMB4ADDRH] = 0x5c,
+	[AMnMB5ADDRL] = 0x60,
+	[AMnMB5ADDRH] = 0x64,
+	[AMnMB6ADDRL] = 0x68,
+	[AMnMB6ADDRH] = 0x6C,
+	[AMnMB7ADDRL] = 0x70,
+	[AMnMB7ADDRH] = 0x74,
+	[AMnMB8ADDRL] = 0x78,
+	[AMnMB8ADDRH] = 0x7C,
+	[AMnMBVALID] = 0x88,
+	[AMnMADRSL] = 0x8C,
+	[AMnMADRSH] = 0x90,
+	[AMnFIFO] = 0x0F0,
+	[AMnFIFOPNTR] = 0xF8,
+	[AMnAXIATTR] = 0x0EC,
+	[AMnAXISTP] = 0x110,
+	[AMnAXISTPACK] = 0x114,
+	[AMnIS] = 0x128,
+	[ICnEN] = 0x1F0,
+	[ICnSVCNUM] = 0x1F8,
+	[ICnSVC] = 0x1FC,
+	[ICnIPMC_C0] = 0x200,
+	[ICnMS] = 0x2D8,
+	[ICnDMR] = 0x304,
+	[ICnTICTRL1] = 0x35C,
+	[ICnTICTRL2] = 0x360,
+	[ICnTISIZE1] = 0x364,
+	[ICnTISIZE2] = 0x368,
+};
+
+static const struct rzg2l_cru_info rzg2l_cru_info = {
+	.cru_type = RZG2L_CRU_TYPE,
+	.regs = rzg2l_cru_regs,
+	.max_width = 2800,
+	.max_height = 4095,
+	.max_cru_channels = 1,
+};
+
+static const struct rzg2l_cru_info rzv2h_cru_info = {
+	.cru_type = RZV2H_CRU_TYPE,
+	.regs = rzv2h_cru_regs,
+	.max_width = 4095,
+	.max_height = 4095,
+	.max_cru_channels = 4,
+};
+
 static const struct of_device_id rzg2l_cru_of_id_table[] = {
-	{ .compatible = "renesas,rzg2l-cru", },
+	{
+		.compatible = "renesas,rzg2l-cru",
+		.data = &rzg2l_cru_info,
+	},
+	{
+		.compatible = "renesas,rzv2h-cru",
+		.data = &rzv2h_cru_info,
+	},
 	{ /* sentinel */ }
 };
 MODULE_DEVICE_TABLE(of, rzg2l_cru_of_id_table);
diff --git a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-cru.h b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-cru.h
index 427f6c243..95823e487 100644
--- a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-cru.h
+++ b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-cru.h
@@ -16,6 +16,61 @@
 #include <media/v4l2-device.h>
 #include <media/videobuf2-v4l2.h>
 
+enum rzg2l_cru_common_regs {
+	CRUnCTRL,	/* CRU Control */
+	CRUnIE,		/* CRU Interrupt Enable (1) */
+	CRUnIE2,	/* CRU Interrupt Enable (2) */
+	CRUnINTS,	/* CRU Interrupt Status (1) */
+	CRUnINTS2,	/* CRU Interrupt Status (2) */
+	CRUnRST,	/* CRU Reset */
+	AMnMB1ADDRL,	/* Bank 1 Address (Lower) for CRU Image Data */
+	AMnMB1ADDRH,	/* Bank 1 Address (Higher) for CRU Image Data */
+	AMnMB2ADDRL,    /* Bank 2 Address (Lower) for CRU Image Data */
+	AMnMB2ADDRH,    /* Bank 2 Address (Higher) for CRU Image Data */
+	AMnMB3ADDRL,    /* Bank 3 Address (Lower) for CRU Image Data */
+	AMnMB3ADDRH,    /* Bank 3 Address (Higher) for CRU Image Data */
+	AMnMB4ADDRL,    /* Bank 4 Address (Lower) for CRU Image Data */
+	AMnMB4ADDRH,    /* Bank 4 Address (Higher) for CRU Image Data */
+	AMnMB5ADDRL,    /* Bank 5 Address (Lower) for CRU Image Data */
+	AMnMB5ADDRH,    /* Bank 5 Address (Higher) for CRU Image Data */
+	AMnMB6ADDRL,    /* Bank 6 Address (Lower) for CRU Image Data */
+	AMnMB6ADDRH,    /* Bank 6 Address (Higher) for CRU Image Data */
+	AMnMB7ADDRL,    /* Bank 7 Address (Lower) for CRU Image Data */
+	AMnMB7ADDRH,    /* Bank 7 Address (Higher) for CRU Image Data */
+	AMnMB8ADDRL,    /* Bank 8 Address (Lower) for CRU Image Data */
+	AMnMB8ADDRH,    /* Bank 8 Address (Higher) for CRU Image Data */
+	AMnMBVALID,	/* Memory Bank Enable for CRU Image Data */
+	AMnMBS,		/* Memory Bank Status for CRU Image Data */
+	AMnMADRSL,	/* VD Memory Address Lower Status Register */
+	AMnMADRSH,	/* VD Memory Address Higher Status Register */
+	AMnAXIATTR,	/* AXI-VD Bus Master Transfer Setting Register */
+	AMnFIFO,        /* AXI-VD Master FIFO Setting Register */
+	AMnFIFOPNTR,	/* AXI Master FIFO Pointer for CRU Image Data */
+	AMnAXISTP,	/* AXI Master Transfer Stop for CRU Image Data */
+	AMnAXISTPACK,	/* AXI Master Transfer Stop Status for CRU Image Data */
+	AMnIS,		/* Image Stride Setting Register */
+	ICnEN,		/* CRU Image Processing Enable */
+	ICnSVCNUM,	/* CRU SVC Number Register */
+	ICnSVC,		/* CRU VC Select Register */
+	ICnMC,		/* CRU Image Processing Main Control */
+	ICnIPMC_C0,	/* CRU Image Converter Main Control 0 */
+	ICnMS,		/* CRU Module Status */
+	ICnDMR,		/* CRU Data Output Mode */
+	ICnTICTRL1,	/* CRU Test Image Generation Control 1 Register */
+	ICnTICTRL2,	/* CRU Test Image Generation Control 2 Register */
+	ICnTISIZE1,	/* CRU Test Image Size Setting 1 Register */
+	ICnTISIZE2,	/* CRU Test Image Size Setting 2 Register */
+
+	CRU_REGS_END,
+};
+
+enum rz_cru_type {
+	RZG2L_CRU_TYPE,
+	RZV2H_CRU_TYPE,
+};
+
+#define RZG2L_CRU_MAX			4
+
 /* Number of HW buffers */
 #define RZG2L_CRU_HW_BUFFER_MAX		8
 #define RZG2L_CRU_HW_BUFFER_DEFAULT	3
@@ -27,9 +82,7 @@
 #define RZG2L_CRU_CSI2_VCHANNEL		4
 
 #define RZG2L_CRU_MIN_INPUT_WIDTH	320
-#define RZG2L_CRU_MAX_INPUT_WIDTH	2800
 #define RZG2L_CRU_MIN_INPUT_HEIGHT	240
-#define RZG2L_CRU_MAX_INPUT_HEIGHT	4095
 
 /**
  * enum rzg2l_cru_dma_state - DMA states
@@ -58,6 +111,15 @@ struct rzg2l_cru_ip {
 	struct v4l2_subdev *remote;
 };
 
+struct rzg2l_cru_info {
+	const u16 *regs;
+	u8 cru_type;
+	unsigned int max_width;
+	unsigned int max_height;
+	int max_cru_channels;
+};
+
+
 /**
  * struct rzg2l_cru_dev - Renesas CRU device structure
  * @dev:		(OF) device
@@ -74,6 +136,7 @@ struct rzg2l_cru_ip {
  * @vdev:		V4L2 video device associated with CRU
  * @v4l2_dev:		V4L2 device
  * @num_buf:		Holds the current number of buffers enabled
+ * @svc_channel:	SVC0/1/2/3 to use
  * @notifier:		V4L2 asynchronous subdevs notifier
  *
  * @ctrl_handler:	V4L2 control handler associated with CRU
@@ -114,6 +177,8 @@ struct rzg2l_cru_dev {
 	struct v4l2_device v4l2_dev;
 	u8 num_buf;
 
+	u8 svc_channel;
+
 	struct v4l2_async_notifier notifier;
 
 	struct v4l2_ctrl_handler ctrl_handler;
@@ -136,6 +201,8 @@ struct rzg2l_cru_dev {
 	enum rzg2l_cru_dma_state state;
 
 	struct v4l2_pix_format format;
+
+	int id;
 };
 
 int rzg2l_cru_start_image_processing(struct rzg2l_cru_dev *cru);
diff --git a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-csi2.c b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-csi2.c
index c7fdee347..1dc6be2c4 100644
--- a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-csi2.c
+++ b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-csi2.c
@@ -85,6 +85,15 @@
 					 CSIDPHYSKW0_UTIL_DL2_SKW_ADJ(1) | \
 					 CSIDPHYSKW0_UTIL_DL3_SKW_ADJ(1))
 
+/* DPHY registers on RZ/V2H(P) SoC */
+#define CRUm_S_TIMCTL			0x41C
+#define CRUm_S_TIMCTL_S_HSSETTLECTL(x)	((x) << 8)
+
+#define CRUm_S_DPHYCTL_MSB		0x434
+#define CRUm_S_DPHYCTL_MSB_DESKEW	BIT(1)
+
+#define CRUm_SWAPCTL			0x438
+
 #define VSRSTS_RETRIES			20
 
 #define RZG2L_CSI2_MIN_WIDTH		320
@@ -107,6 +116,7 @@ struct rzg2l_csi2 {
 	void __iomem *base;
 	struct reset_control *presetn;
 	struct reset_control *cmn_rstb;
+	const struct rzg2l_csi2_info *info;
 	struct clk *sysclk;
 	struct clk *vclk;
 	unsigned long vclk_rate;
@@ -123,6 +133,11 @@ struct rzg2l_csi2 {
 	bool dphy_enabled;
 };
 
+struct rzg2l_csi2_info {
+	int (*dphy_enable)(struct rzg2l_csi2 *csi2);
+	int (*dphy_disable)(struct rzg2l_csi2 *csi2);
+};
+
 struct rzg2l_csi2_timings {
 	u32 t_init;
 	u32 tclk_miss;
@@ -133,6 +148,30 @@ struct rzg2l_csi2_timings {
 	u32 max_hsfreq;
 };
 
+struct rzv2h_csi2_s_hssettlectl {
+	unsigned int hsfreq;
+	u16 s_hssettlectl;
+};
+
+static const struct rzv2h_csi2_s_hssettlectl rzv2h_s_hssettlectl[] = {
+	{   90,  1 }, {  130,  2 }, {  180,  3 },
+	{  220,  4 }, {  270,  5 }, {  310,  6 },
+	{  360,  7 }, {  400,  8 }, {  450,  9 },
+	{  490, 10 }, {  540, 11 }, {  580, 12 },
+	{  630, 13 }, {  670, 14 }, {  720, 15 },
+	{  760, 16 }, {  810, 17 }, {  850, 18 },
+	{  900, 19 }, {  940, 20 }, {  990, 21 },
+	{ 1030, 22 }, { 1080, 23 }, { 1120, 24 },
+	{ 1170, 25 }, { 1220, 26 }, { 1260, 27 },
+	{ 1310, 28 }, { 1350, 29 }, { 1400, 30 },
+	{ 1440, 31 }, { 1490, 32 }, { 1530, 33 },
+	{ 1580, 34 }, { 1620, 35 }, { 1670, 36 },
+	{ 1710, 37 }, { 1760, 38 }, { 1800, 39 },
+	{ 1850, 40 }, { 1890, 41 }, { 1940, 42 },
+	{ 1980, 43 }, { 2030, 44 }, { 2070, 45 },
+	{ 2100, 46 },
+};
+
 static const struct rzg2l_csi2_timings rzg2l_csi2_global_timings[] = {
 	{
 		.max_hsfreq = 80,
@@ -352,14 +391,63 @@ static int rzg2l_csi2_dphy_enable(struct rzg2l_csi2 *csi2)
 	return ret;
 }
 
+static int rzv2h_csi2_dphy_disable(struct rzg2l_csi2 *csi2)
+{
+ 	int ret;
+ 
+ 	/* Reset the CRU (D-PHY) */
+ 	ret = reset_control_assert(csi2->cmn_rstb);
+ 	if (ret)
+ 		return ret;
+ 
+ 	csi2->dphy_enabled = false;
+ 
+ 	return 0;
+}
+ 
+static int rzv2h_csi2_dphy_enable(struct rzg2l_csi2 *csi2)
+{
+ 	unsigned int i;
+ 	u16 hssettle;
+ 	int mbps;
+ 
+ 	mbps = rzg2l_csi2_calc_mbps(csi2);
+ 	if (mbps < 0)
+ 		return mbps;
+ 
+ 	csi2->hsfreq = mbps;
+ 
+ 	rzg2l_csi2_write(csi2, CRUm_SWAPCTL, 0);
+ 
+ 	for (i = 0; i < ARRAY_SIZE(rzv2h_s_hssettlectl); i++) {
+ 		if (csi2->hsfreq <= rzv2h_s_hssettlectl[i].hsfreq)
+ 			break;
+ 	}
+ 
+ 	if (i == ARRAY_SIZE(rzv2h_s_hssettlectl))
+ 		return -EINVAL;
+ 
+ 	hssettle = rzv2h_s_hssettlectl[i].s_hssettlectl;
+ 	rzg2l_csi2_write(csi2, CRUm_S_TIMCTL, CRUm_S_TIMCTL_S_HSSETTLECTL(hssettle));
+ 
+ 	if (csi2->hsfreq > 1500)
+ 		rzg2l_csi2_set(csi2, CRUm_S_DPHYCTL_MSB, CRUm_S_DPHYCTL_MSB_DESKEW);
+ 	else
+ 		rzg2l_csi2_clr(csi2, CRUm_S_DPHYCTL_MSB, CRUm_S_DPHYCTL_MSB_DESKEW);
+ 
+ 	csi2->dphy_enabled = true;
+ 
+ 	return 0;
+}
+
 static int rzg2l_csi2_dphy_setting(struct v4l2_subdev *sd, bool on)
 {
 	struct rzg2l_csi2 *csi2 = sd_to_csi2(sd);
 
 	if (on)
-		return rzg2l_csi2_dphy_enable(csi2);
+		return csi2->info->dphy_enable(csi2);
 
-	return rzg2l_csi2_dphy_disable(csi2);
+	return csi2->info->dphy_disable(csi2);
 }
 
 static int rzg2l_csi2_mipi_link_enable(struct rzg2l_csi2 *csi2)
@@ -754,12 +842,12 @@ static int rzg2l_csi2_probe(struct platform_device *pdev)
 		return dev_err_probe(&pdev->dev, PTR_ERR(csi2->cmn_rstb),
 				     "Failed to get cpg cmn-rstb\n");
 
-	csi2->presetn = devm_reset_control_get_shared(&pdev->dev, "presetn");
+	csi2->presetn = devm_reset_control_get_optional_shared(&pdev->dev, "presetn");
 	if (IS_ERR(csi2->presetn))
 		return dev_err_probe(&pdev->dev, PTR_ERR(csi2->presetn),
 				     "Failed to get cpg presetn\n");
 
-	csi2->sysclk = devm_clk_get(&pdev->dev, "system");
+	csi2->sysclk = devm_clk_get_optional(&pdev->dev, "system");
 	if (IS_ERR(csi2->sysclk))
 		return dev_err_probe(&pdev->dev, PTR_ERR(csi2->sysclk),
 				     "Failed to get system clk\n");
@@ -772,6 +860,8 @@ static int rzg2l_csi2_probe(struct platform_device *pdev)
 
 	csi2->dev = &pdev->dev;
 
+	csi2->info = of_device_get_match_data(&pdev->dev);
+
 	platform_set_drvdata(pdev, csi2);
 
 	ret = rzg2l_csi2_parse_dt(csi2);
@@ -861,8 +951,25 @@ static const struct dev_pm_ops rzg2l_csi2_pm_ops = {
 		       rzg2l_csi2_pm_runtime_resume, NULL)
 };
 
+static const struct rzg2l_csi2_info rzv2h_csi2_info = {
+	.dphy_enable = rzv2h_csi2_dphy_enable,
+	.dphy_disable = rzv2h_csi2_dphy_disable,
+};
+
+static const struct rzg2l_csi2_info rzg2l_csi2_info = {
+	.dphy_enable = rzg2l_csi2_dphy_enable,
+	.dphy_disable = rzg2l_csi2_dphy_disable,
+};
+
 static const struct of_device_id rzg2l_csi2_of_table[] = {
-	{ .compatible = "renesas,rzg2l-csi2", },
+	{
+		.compatible = "renesas,rzg2l-csi2",
+		.data = &rzg2l_csi2_info,
+	},
+	{
+		.compatible = "renesas,rzv2h-csi2",
+		.data = &rzv2h_csi2_info,
+	},
 	{ /* sentinel */ }
 };
 MODULE_DEVICE_TABLE(of, rzg2l_csi2_of_table);
diff --git a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-ip.c b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-ip.c
index ac8ebae4e..7fc3cdbd5 100644
--- a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-ip.c
+++ b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-ip.c
@@ -97,6 +97,8 @@ static int rzg2l_cru_ip_set_format(struct v4l2_subdev *sd,
 				   struct v4l2_subdev_state *state,
 				   struct v4l2_subdev_format *fmt)
 {
+	struct rzg2l_cru_dev *cru = v4l2_get_subdevdata(sd);
+	const struct rzg2l_cru_info *info = cru->info;
 	struct v4l2_mbus_framefmt *src_format;
 	struct v4l2_mbus_framefmt *sink_format;
 
@@ -119,9 +121,9 @@ static int rzg2l_cru_ip_set_format(struct v4l2_subdev *sd,
 	sink_format->ycbcr_enc = fmt->format.ycbcr_enc;
 	sink_format->quantization = fmt->format.quantization;
 	sink_format->width = clamp_t(u32, fmt->format.width,
-				     RZG2L_CRU_MIN_INPUT_WIDTH, RZG2L_CRU_MAX_INPUT_WIDTH);
+				     RZG2L_CRU_MIN_INPUT_WIDTH, info->max_width);
 	sink_format->height = clamp_t(u32, fmt->format.height,
-				      RZG2L_CRU_MIN_INPUT_HEIGHT, RZG2L_CRU_MAX_INPUT_HEIGHT);
+				      RZG2L_CRU_MIN_INPUT_HEIGHT, info->max_height);
 
 	fmt->format = *sink_format;
 
@@ -146,6 +148,9 @@ static int rzg2l_cru_ip_enum_frame_size(struct v4l2_subdev *sd,
 					struct v4l2_subdev_state *state,
 					struct v4l2_subdev_frame_size_enum *fse)
 {
+	struct rzg2l_cru_dev *cru = v4l2_get_subdevdata(sd);
+	const struct rzg2l_cru_info *info = cru->info;
+	
 	if (fse->index != 0)
 		return -EINVAL;
 
@@ -154,8 +159,8 @@ static int rzg2l_cru_ip_enum_frame_size(struct v4l2_subdev *sd,
 
 	fse->min_width = RZG2L_CRU_MIN_INPUT_WIDTH;
 	fse->min_height = RZG2L_CRU_MIN_INPUT_HEIGHT;
-	fse->max_width = RZG2L_CRU_MAX_INPUT_WIDTH;
-	fse->max_height = RZG2L_CRU_MAX_INPUT_HEIGHT;
+	fse->max_width = info->max_width;
+	fse->max_height = info->max_height;
 
 	return 0;
 }
diff --git a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-video.c b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-video.c
index b16b8af6e..5ac950a42 100644
--- a/drivers/media/platform/renesas/rzg2l-cru/rzg2l-video.c
+++ b/drivers/media/platform/renesas/rzg2l-cru/rzg2l-video.c
@@ -1,15 +1,15 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
- * Driver for Renesas RZ/G2L CRU
- *
- * Copyright (C) 2022 Renesas Electronics Corp.
- *
- * Based on Renesas R-Car VIN
- * Copyright (C) 2016 Renesas Electronics Corp.
- * Copyright (C) 2011-2013 Renesas Solutions Corp.
- * Copyright (C) 2013 Cogent Embedded, Inc., <source@cogentembedded.com>
- * Copyright (C) 2008 Magnus Damm
- */
+* Driver for Renesas RZ/G2L CRU
+*
+* Copyright (C) 2022 Renesas Electronics Corp.
+*
+* Based on Renesas R-Car VIN
+* Copyright (C) 2016 Renesas Electronics Corp.
+* Copyright (C) 2011-2013 Renesas Solutions Corp.
+* Copyright (C) 2013 Cogent Embedded, Inc., <source@cogentembedded.com>
+* Copyright (C) 2008 Magnus Damm
+*/
 
 #include <linux/clk.h>
 #include <linux/delay.h>
@@ -20,71 +20,79 @@
 
 #include "rzg2l-cru.h"
 
-/* HW CRU Registers Definition */
-
-/* CRU Control Register */
-#define CRUnCTRL			0x0
 #define CRUnCTRL_VINSEL(x)		((x) << 0)
 
-/* CRU Interrupt Enable Register */
-#define CRUnIE				0x4
 #define CRUnIE_EFE			BIT(17)
-
-/* CRU Interrupt Status Register */
-#define CRUnINTS			0x8
+#define CRUnIE2_FEE(x)			BIT(((x) * 3) + 1)
 #define CRUnINTS_SFS			BIT(16)
 
-/* CRU Reset Register */
-#define CRUnRST				0xc
+#define CRUnINTS2_FSS(x)		BIT((x) * 3)
+
 #define CRUnRST_VRESETN			BIT(0)
 
 /* Memory Bank Base Address (Lower) Register for CRU Image Data */
-#define AMnMBxADDRL(x)			(0x100 + ((x) * 8))
+#define AMnMBxADDRL(base, x)		((base) + (x) * 2)
 
 /* Memory Bank Base Address (Higher) Register for CRU Image Data */
-#define AMnMBxADDRH(x)			(0x104 + ((x) * 8))
+#define AMnMBxADDRH(base, x)		((base) + (x) * 2)
 
-/* Memory Bank Enable Register for CRU Image Data */
-#define AMnMBVALID			0x148
 #define AMnMBVALID_MBVALID(x)		GENMASK(x, 0)
 
-/* Memory Bank Status Register for CRU Image Data */
-#define AMnMBS				0x14c
 #define AMnMBS_MBSTS			0x7
 
-/* AXI Master FIFO Pointer Register for CRU Image Data */
-#define AMnFIFOPNTR			0x168
+/* AXI-VD Bus Master Transfer Setting Register */
+#define AMnAXIATTR_AXILEN_MASK		GENMASK(3, 0)
+#define AMnAXIATTR_AXILEN		(0xf)
+
 #define AMnFIFOPNTR_FIFOWPNTR		GENMASK(7, 0)
+#define AMnFIFOPNTR_FIFOWPNTR_B0	AMnFIFOPNTR_FIFOWPNTR
+#define AMnFIFOPNTR_FIFOWPNTR_B1	GENMASK(15, 8)
 #define AMnFIFOPNTR_FIFORPNTR_Y		GENMASK(23, 16)
+#define AMnFIFOPNTR_FIFORPNTR_B0	AMnFIFOPNTR_FIFORPNTR_Y
+#define AMnFIFOPNTR_FIFORPNTR_B1	GENMASK(31, 24)
+
+#define AMnIS_IS(x)			((x) << 7)
 
-/* AXI Master Transfer Stop Register for CRU Image Data */
-#define AMnAXISTP			0x174
 #define AMnAXISTP_AXI_STOP		BIT(0)
 
-/* AXI Master Transfer Stop Status Register for CRU Image Data */
-#define AMnAXISTPACK			0x178
 #define AMnAXISTPACK_AXI_STOP_ACK	BIT(0)
 
-/* CRU Image Processing Enable Register */
-#define ICnEN				0x200
 #define ICnEN_ICEN			BIT(0)
 
-/* CRU Image Processing Main Control Register */
-#define ICnMC				0x208
+#define ICnSVC_SVC0(x)			(x)
+#define ICnSVC_SVC1(x)			((x) << 4)
+#define ICnSVC_SVC2(x)			((x) << 8)
+#define ICnSVC_SVC3(x)			((x) << 12)
+
 #define ICnMC_CSCTHR			BIT(5)
 #define ICnMC_INF_YUV8_422		(0x1e << 16)
 #define ICnMC_INF_USER			(0x30 << 16)
 #define ICnMC_VCSEL(x)			((x) << 22)
 #define ICnMC_INF_MASK			GENMASK(21, 16)
 
-/* CRU Module Status Register */
-#define ICnMS				0x254
 #define ICnMS_IA			BIT(2)
 
-/* CRU Data Output Mode Register */
-#define ICnDMR				0x26c
 #define ICnDMR_YCMODE_UYVY		(1 << 4)
 
+/* CRU Test Image Generation Control 1 Register */
+#define ICnTICTRL1_TIEN			BIT(0)
+#define ICnTICTRL1_TIMODE		BIT(1)
+#define ICnTICTRL1_TIPTNY1(x)		((x) << 4)
+#define ICnTICTRL1_TIPTNU1(x)		((x) << 8)
+#define ICnTICTRL1_TIPTNV1(x)		((x) << 12)
+
+/* CRU Test Image Generation Control 2 Register */
+#define ICnTICTRL2_TIPTNY2(x)		((x) << 0)
+#define ICnTICTRL2_TIPTNU2(x)		((x) << 8)
+#define ICnTICTRL2_TIPTNV2(x)		((x) << 16)
+
+/* CRU Test Image Size Setting 1 Register */
+#define ICnTISIZE1_TIPPL(x)		((x) << 0)
+
+/* CRU Test Image Size Setting 2 Register */
+#define ICnTISIZE2_TIN(x)		((x) << 0)
+#define ICnTISIZE2_TIM(x)		((x) << 16)
+
 #define RZG2L_TIMEOUT_MS		100
 #define RZG2L_RETRIES			10
 
@@ -99,25 +107,47 @@ struct rzg2l_cru_buffer {
 	struct list_head list;
 };
 
+static int prev_slot[RZG2L_CRU_MAX];
+static u32 amnmbxaddrl[RZG2L_CRU_MAX][RZG2L_CRU_HW_BUFFER_MAX];
+static u32 amnmbxaddrh[RZG2L_CRU_MAX][RZG2L_CRU_HW_BUFFER_MAX];
+
 #define to_buf_list(vb2_buffer) \
 	(&container_of(vb2_buffer, struct rzg2l_cru_buffer, vb)->list)
 
 /* -----------------------------------------------------------------------------
- * DMA operations
- */
+* DMA operations
+*/
 static void rzg2l_cru_write(struct rzg2l_cru_dev *cru, u32 offset, u32 value)
 {
-	iowrite32(value, cru->base + offset);
+	const u16 *regs = cru->info->regs;
+
+	iowrite32(value, cru->base + regs[offset]);
 }
 
 static u32 rzg2l_cru_read(struct rzg2l_cru_dev *cru, u32 offset)
 {
-	return ioread32(cru->base + offset);
+	const u16 *regs = cru->info->regs;
+
+	return ioread32(cru->base + regs[offset]);
+}
+
+static void rzg2l_cru_set_mb(struct rzg2l_cru_dev *cru,
+				u32 slot, dma_addr_t addr)
+{
+	rzg2l_cru_write(cru, AMnMBxADDRL(AMnMB1ADDRL, slot), lower_32_bits(addr));
+	rzg2l_cru_write(cru, AMnMBxADDRH(AMnMB1ADDRH, slot), upper_32_bits(addr));
+}
+
+static void rzg2l_cru_get_mb(struct rzg2l_cru_dev *cru,
+				u32 slot, u32 *low, u32 *high)
+{
+	*low = rzg2l_cru_read(cru, AMnMBxADDRL(AMnMB1ADDRL, slot));
+	*high = rzg2l_cru_read(cru, AMnMBxADDRH(AMnMB1ADDRH, slot));
 }
 
 /* Need to hold qlock before calling */
 static void return_unused_buffers(struct rzg2l_cru_dev *cru,
-				  enum vb2_buffer_state state)
+				enum vb2_buffer_state state)
 {
 	struct rzg2l_cru_buffer *buf, *node;
 	unsigned long flags;
@@ -140,8 +170,8 @@ static void return_unused_buffers(struct rzg2l_cru_dev *cru,
 }
 
 static int rzg2l_cru_queue_setup(struct vb2_queue *vq, unsigned int *nbuffers,
-				 unsigned int *nplanes, unsigned int sizes[],
-				 struct device *alloc_devs[])
+				unsigned int *nplanes, unsigned int sizes[],
+				struct device *alloc_devs[])
 {
 	struct rzg2l_cru_dev *cru = vb2_get_drv_priv(vq);
 
@@ -218,33 +248,34 @@ static int rzg2l_cru_mc_validate_format(struct rzg2l_cru_dev *cru,
 	}
 
 	if (fmt.format.width != cru->format.width ||
-	    fmt.format.height != cru->format.height)
+		fmt.format.height != cru->format.height)
 		return -EPIPE;
 
 	return 0;
 }
 
 static void rzg2l_cru_set_slot_addr(struct rzg2l_cru_dev *cru,
-				    int slot, dma_addr_t addr)
+					int slot, dma_addr_t addr)
 {
 	/*
-	 * The address needs to be 512 bytes aligned. Driver should never accept
-	 * settings that do not satisfy this in the first place...
-	 */
+	* The address needs to be 512 bytes aligned. Driver should never accept
+	* settings that do not satisfy this in the first place...
+	*/
 	if (WARN_ON((addr) & RZG2L_CRU_HW_BUFFER_MASK))
 		return;
 
-	/* Currently, we just use the buffer in 32 bits address */
-	rzg2l_cru_write(cru, AMnMBxADDRL(slot), addr);
-	rzg2l_cru_write(cru, AMnMBxADDRH(slot), 0);
+	rzg2l_cru_set_mb(cru, slot, addr);
+
+	amnmbxaddrl[cru->id][slot] = lower_32_bits(addr);
+	amnmbxaddrh[cru->id][slot] = upper_32_bits(addr);
 }
 
 /*
- * Moves a buffer from the queue to the HW slot. If no buffer is
- * available use the scratch buffer. The scratch buffer is never
- * returned to userspace, its only function is to enable the capture
- * loop to keep running.
- */
+* Moves a buffer from the queue to the HW slot. If no buffer is
+* available use the scratch buffer. The scratch buffer is never
+* returned to userspace, its only function is to enable the capture
+* loop to keep running.
+*/
 static void rzg2l_cru_fill_hw_slot(struct rzg2l_cru_dev *cru, int slot)
 {
 	struct vb2_v4l2_buffer *vbuf;
@@ -263,7 +294,7 @@ static void rzg2l_cru_fill_hw_slot(struct rzg2l_cru_dev *cru, int slot)
 	} else {
 		/* Keep track of buffer we give to HW */
 		buf = list_entry(cru->buf_list.next,
-				 struct rzg2l_cru_buffer, list);
+				struct rzg2l_cru_buffer, list);
 		vbuf = &buf->vb;
 		list_del_init(to_buf_list(vbuf));
 		cru->queue_buf[slot] = vbuf;
@@ -273,24 +304,32 @@ static void rzg2l_cru_fill_hw_slot(struct rzg2l_cru_dev *cru, int slot)
 	}
 
 	rzg2l_cru_set_slot_addr(cru, slot, phys_addr);
+	rzg2l_cru_get_mb(cru, slot, &amnmbxaddrl[cru->id][slot],
+			&amnmbxaddrh[cru->id][slot]);
 }
 
 static void rzg2l_cru_initialize_axi(struct rzg2l_cru_dev *cru)
 {
 	unsigned int slot;
+	u32 amnaxiattr;
 
 	/*
-	 * Set image data memory banks.
-	 * Currently, we will use maximum address.
-	 */
+	* Set image data memory banks.
+	* Currently, we will use maximum address.
+	*/
 	rzg2l_cru_write(cru, AMnMBVALID, AMnMBVALID_MBVALID(cru->num_buf - 1));
 
 	for (slot = 0; slot < cru->num_buf; slot++)
 		rzg2l_cru_fill_hw_slot(cru, slot);
+
+	/* Set AXI burst max length to recommended setting */
+	amnaxiattr = rzg2l_cru_read(cru, AMnAXIATTR) & ~AMnAXIATTR_AXILEN_MASK;
+	amnaxiattr |= AMnAXIATTR_AXILEN;
+	rzg2l_cru_write(cru, AMnAXIATTR, amnaxiattr);
 }
 
 static void rzg2l_cru_csi2_setup(struct rzg2l_cru_dev *cru, bool *input_is_yuv,
-				 struct v4l2_mbus_framefmt *ip_sd_fmt)
+				struct v4l2_mbus_framefmt *ip_sd_fmt)
 {
 	u32 icnmc;
 
@@ -305,19 +344,27 @@ static void rzg2l_cru_csi2_setup(struct rzg2l_cru_dev *cru, bool *input_is_yuv,
 		break;
 	}
 
-	icnmc |= (rzg2l_cru_read(cru, ICnMC) & ~ICnMC_INF_MASK);
-
-	/* Set virtual channel CSI2 */
-	icnmc |= ICnMC_VCSEL(cru->csi.channel);
+	if (cru->info->cru_type == RZG2L_CRU_TYPE) {
+		icnmc |= (rzg2l_cru_read(cru, ICnMC) & ~ICnMC_INF_MASK);
+		/* Set virtual channel CSI2 */
+		icnmc |= ICnMC_VCSEL(cru->csi.channel);
+		rzg2l_cru_write(cru, ICnMC, icnmc);
+	} else {
+		icnmc |= (rzg2l_cru_read(cru, ICnIPMC_C0) & ~ICnMC_INF_MASK);
 
-	rzg2l_cru_write(cru, ICnMC, icnmc);
+		rzg2l_cru_write(cru, ICnSVCNUM, cru->svc_channel);
+		rzg2l_cru_write(cru, ICnSVC, ICnSVC_SVC0(0) | ICnSVC_SVC1(1) |
+				ICnSVC_SVC2(2) | ICnSVC_SVC3(3));
+		rzg2l_cru_write(cru, ICnIPMC_C0, icnmc);
+	}
 }
 
 static int rzg2l_cru_initialize_image_conv(struct rzg2l_cru_dev *cru,
-					   struct v4l2_mbus_framefmt *ip_sd_fmt)
+					struct v4l2_mbus_framefmt *ip_sd_fmt)
 {
 	bool output_is_yuv = false;
 	bool input_is_yuv = false;
+	u32 icmc_reg = ICnMC;
 	u32 icndmr;
 
 	rzg2l_cru_csi2_setup(cru, &input_is_yuv, ip_sd_fmt);
@@ -334,13 +381,15 @@ static int rzg2l_cru_initialize_image_conv(struct rzg2l_cru_dev *cru,
 		return -EINVAL;
 	}
 
+	icmc_reg = (cru->info->cru_type == RZG2L_CRU_TYPE) ? ICnMC : ICnIPMC_C0;
+
 	/* If input and output use same colorspace, do bypass mode */
 	if (output_is_yuv == input_is_yuv)
-		rzg2l_cru_write(cru, ICnMC,
-				rzg2l_cru_read(cru, ICnMC) | ICnMC_CSCTHR);
+		rzg2l_cru_write(cru, icmc_reg,
+				rzg2l_cru_read(cru, icmc_reg) | ICnMC_CSCTHR);
 	else
-		rzg2l_cru_write(cru, ICnMC,
-				rzg2l_cru_read(cru, ICnMC) & (~ICnMC_CSCTHR));
+		rzg2l_cru_write(cru, icmc_reg,
+				rzg2l_cru_read(cru, icmc_reg) & (~ICnMC_CSCTHR));
 
 	/* Set output data format */
 	rzg2l_cru_write(cru, ICnDMR, icndmr);
@@ -348,6 +397,29 @@ static int rzg2l_cru_initialize_image_conv(struct rzg2l_cru_dev *cru,
 	return 0;
 }
 
+static void rzg2l_cru_disable_interrupts(struct rzg2l_cru_dev *cru)
+{
+	/* Disable and clear the interrupt before using */
+	if (cru->info->cru_type == RZG2L_CRU_TYPE) {
+		rzg2l_cru_write(cru, CRUnIE, 0);
+		rzg2l_cru_write(cru, CRUnINTS, 0x001F000F);
+	} else {
+		rzg2l_cru_write(cru, CRUnIE, 0);
+		rzg2l_cru_write(cru, CRUnIE2, 0);
+		rzg2l_cru_write(cru, CRUnINTS, 0x80000F0F);
+		rzg2l_cru_write(cru, CRUnINTS2, 0x31F0FFF);
+	}
+}
+
+static void rzg2l_cru_enable_interrupts(struct rzg2l_cru_dev *cru)
+{
+	/* Enable interrupts */
+	if (cru->info->cru_type == RZG2L_CRU_TYPE)
+		rzg2l_cru_write(cru, CRUnIE, CRUnIE_EFE);
+	else
+		rzg2l_cru_write(cru, CRUnIE2, CRUnIE2_FEE(cru->svc_channel));
+}
+
 void rzg2l_cru_stop_image_processing(struct rzg2l_cru_dev *cru)
 {
 	u32 amnfifopntr, amnfifopntr_w, amnfifopntr_r_y;
@@ -358,8 +430,7 @@ void rzg2l_cru_stop_image_processing(struct rzg2l_cru_dev *cru)
 	spin_lock_irqsave(&cru->qlock, flags);
 
 	/* Disable and clear the interrupt */
-	rzg2l_cru_write(cru, CRUnIE, 0);
-	rzg2l_cru_write(cru, CRUnINTS, 0x001F0F0F);
+	rzg2l_cru_disable_interrupts(cru);
 
 	/* Stop the operation of image conversion */
 	rzg2l_cru_write(cru, ICnEN, 0);
@@ -381,11 +452,19 @@ void rzg2l_cru_stop_image_processing(struct rzg2l_cru_dev *cru)
 	for (retries = 5; retries > 0; retries--) {
 		amnfifopntr = rzg2l_cru_read(cru, AMnFIFOPNTR);
 
-		amnfifopntr_w = amnfifopntr & AMnFIFOPNTR_FIFOWPNTR;
-		amnfifopntr_r_y =
-			(amnfifopntr & AMnFIFOPNTR_FIFORPNTR_Y) >> 16;
-		if (amnfifopntr_w == amnfifopntr_r_y)
-			break;
+		if (cru->info->cru_type == RZG2L_CRU_TYPE) {
+			amnfifopntr_w = amnfifopntr & AMnFIFOPNTR_FIFOWPNTR;
+			amnfifopntr_r_y =
+				(amnfifopntr & AMnFIFOPNTR_FIFORPNTR_Y) >> 16;
+			if (amnfifopntr_w == amnfifopntr_r_y)
+				break;
+		} else {
+			if ((((amnfifopntr & AMnFIFOPNTR_FIFORPNTR_B1) >> 24) ==
+				((amnfifopntr & AMnFIFOPNTR_FIFOWPNTR_B1) >> 8)) &&
+				(((amnfifopntr & AMnFIFOPNTR_FIFORPNTR_B0) >> 16) ==
+				(amnfifopntr & AMnFIFOPNTR_FIFOWPNTR_B0)))
+				break;
+		}
 
 		usleep_range(10, 20);
 	}
@@ -404,7 +483,7 @@ void rzg2l_cru_stop_image_processing(struct rzg2l_cru_dev *cru)
 			break;
 
 		usleep_range(10, 20);
-	}
+	};
 
 	/* Notify that AXI bus can not stop here */
 	if (!retries)
@@ -426,6 +505,7 @@ int rzg2l_cru_start_image_processing(struct rzg2l_cru_dev *cru)
 {
 	struct v4l2_mbus_framefmt *fmt = rzg2l_cru_ip_get_src_fmt(cru);
 	unsigned long flags;
+	u32 stride;
 	int ret;
 
 	spin_lock_irqsave(&cru->qlock, flags);
@@ -437,12 +517,23 @@ int rzg2l_cru_start_image_processing(struct rzg2l_cru_dev *cru)
 	rzg2l_cru_write(cru, CRUnRST, CRUnRST_VRESETN);
 
 	/* Disable and clear the interrupt before using */
-	rzg2l_cru_write(cru, CRUnIE, 0);
-	rzg2l_cru_write(cru, CRUnINTS, 0x001f000f);
+	rzg2l_cru_disable_interrupts(cru);
 
 	/* Initialize the AXI master */
 	rzg2l_cru_initialize_axi(cru);
 
+	if (cru->info->cru_type != RZG2L_CRU_TYPE) {
+		stride = cru->format.bytesperline;
+		if (stride % 128) {
+			spin_unlock_irqrestore(&cru->qlock, flags);
+			dev_err(cru->dev,
+				"Bytesperline must be multiple of 128 bytes\n");
+			return -EINVAL;
+		}
+		stride = stride / 128;
+		rzg2l_cru_write(cru, AMnIS, AMnIS_IS(stride));
+	}
+
 	/* Initialize image convert */
 	ret = rzg2l_cru_initialize_image_conv(cru, fmt);
 	if (ret) {
@@ -451,7 +542,7 @@ int rzg2l_cru_start_image_processing(struct rzg2l_cru_dev *cru)
 	}
 
 	/* Enable interrupt */
-	rzg2l_cru_write(cru, CRUnIE, CRUnIE_EFE);
+	rzg2l_cru_enable_interrupts(cru);
 
 	/* Enable image processing reception */
 	rzg2l_cru_write(cru, ICnEN, ICnEN_ICEN);
@@ -466,7 +557,8 @@ static int rzg2l_cru_set_stream(struct rzg2l_cru_dev *cru, int on)
 	struct media_pipeline *pipe;
 	struct v4l2_subdev *sd;
 	struct media_pad *pad;
-	int ret;
+	int ret, i;
+	unsigned long flags;
 
 	pad = media_pad_remote_pad_first(&cru->pad);
 	if (!pad)
@@ -505,6 +597,14 @@ static int rzg2l_cru_set_stream(struct rzg2l_cru_dev *cru, int on)
 	if (ret && ret != -ENOIOCTLCMD)
 		goto pipe_line_stop;
 
+	spin_lock_irqsave(&cru->qlock, flags);
+
+	for (i = 0; i < cru->num_buf; i++)
+		rzg2l_cru_get_mb(cru, i, &amnmbxaddrl[cru->id][i],
+				&amnmbxaddrh[cru->id][i]);
+
+	spin_unlock_irqrestore(&cru->qlock, flags);
+
 	ret = v4l2_subdev_call(sd, video, s_stream, 1);
 	if (ret && ret != -ENOIOCTLCMD)
 		goto err_s_stream;
@@ -564,18 +664,18 @@ static irqreturn_t rzg2l_cru_irq(int irq, void *data)
 	slot = amnmbs & AMnMBS_MBSTS;
 
 	/*
-	 * AMnMBS.MBSTS indicates the destination of Memory Bank (MB).
-	 * Recalculate to get the current transfer complete MB.
-	 */
+	* AMnMBS.MBSTS indicates the destination of Memory Bank (MB).
+	* Recalculate to get the current transfer complete MB.
+	*/
 	if (slot == 0)
 		slot = cru->num_buf - 1;
 	else
 		slot--;
 
 	/*
-	 * To hand buffers back in a known order to userspace start
-	 * to capture first from slot 0.
-	 */
+	* To hand buffers back in a known order to userspace start
+	* to capture first from slot 0.
+	*/
 	if (cru->state == RZG2L_CRU_DMA_STARTING) {
 		if (slot != 0) {
 			dev_dbg(cru->dev, "Starting sync slot: %d\n", slot);
@@ -610,6 +710,117 @@ static irqreturn_t rzg2l_cru_irq(int irq, void *data)
 	return IRQ_RETVAL(handled);
 }
 
+static irqreturn_t rzv2h_cru_irq(int irq, void *data)
+{
+	struct rzg2l_cru_dev *cru = data;
+	bool write_complete = false;
+	unsigned int handled = 0;
+	dma_addr_t amnmadrs;
+	unsigned long flags;
+	unsigned int slot;
+	u32 irq_status;
+
+	spin_lock_irqsave(&cru->qlock, flags);
+
+	irq_status = rzg2l_cru_read(cru, CRUnINTS2);
+	if (!(irq_status))
+		goto done;
+
+	handled = 1;
+
+	rzg2l_cru_write(cru, CRUnINTS2, irq_status);
+
+	/* Nothing to do if capture status is 'STOPPED' */
+	if (cru->state == RZG2L_CRU_DMA_STOPPED) {
+		dev_dbg(cru->dev, "IRQ while state stopped\n");
+		goto done;
+	}
+
+	if (cru->state == RZG2L_CRU_DMA_STOPPING) {
+		if (irq_status & (CRUnINTS2_FSS(0) | CRUnINTS2_FSS(1) |
+				CRUnINTS2_FSS(2) | CRUnINTS2_FSS(3)))
+			dev_dbg(cru->dev, "IRQ while state stopping\n");
+		goto done;
+	}
+
+	amnmadrs = rzg2l_cru_read(cru, AMnMADRSL);
+	amnmadrs |= (((unsigned long)rzg2l_cru_read(cru, AMnMADRSH)) << 32);
+
+	/* Check current HW slot based on current MB address */
+	for (slot = 0; slot < cru->num_buf; slot++) {
+		dma_addr_t tmp;
+
+		tmp = amnmbxaddrh[cru->id][slot];
+		tmp = (tmp << 32) | amnmbxaddrl[cru->id][slot];
+
+		tmp = amnmadrs - tmp;
+		if (((long)tmp) && tmp <= cru->format.sizeimage) {
+			write_complete = 1;
+			break;
+		}
+	}
+
+	if (!write_complete) {
+		dev_err(cru->dev, "Invalid MB address 0x%llx\n", amnmadrs);
+		goto done;
+	}
+
+	amnmbxaddrl[cru->id][slot] = 0;
+	amnmbxaddrh[cru->id][slot] = 0;
+
+	/*
+	* To hand buffers back in a known order to userspace start
+	* to capture first from slot 0.
+	*/
+	if (cru->state == RZG2L_CRU_DMA_STARTING) {
+		if (slot != 0) {
+			dev_dbg(cru->dev, "Starting sync slot: %d\n", slot);
+			goto done;
+		}
+
+		dev_dbg(cru->dev, "Capture start synced!\n");
+		cru->state = RZG2L_CRU_DMA_RUNNING;
+	}
+
+	if (slot != prev_slot[cru->id]) {
+		/* Update value of previous memory bank slot */
+		prev_slot[cru->id] = slot;
+	} else {
+		/*
+		* AXI-Bus congestion maybe occurred.
+		* Set auto recovery mode to clear all FIFOs
+		* and resume transmission.
+		*/
+		rzg2l_cru_write(cru, AMnFIFO, 0);
+
+		dev_dbg(cru->dev, "Dropping frame %u with CRU channel %d\n",
+			cru->sequence, cru->id);
+	}
+
+	/* Capture frame */
+	if (cru->queue_buf[slot]) {
+		cru->queue_buf[slot]->field = cru->format.field;
+		cru->queue_buf[slot]->sequence = cru->sequence;
+		cru->queue_buf[slot]->vb2_buf.timestamp = ktime_get_ns();
+		vb2_buffer_done(&cru->queue_buf[slot]->vb2_buf,
+				VB2_BUF_STATE_DONE);
+		cru->queue_buf[slot] = NULL;
+	} else {
+		/* Scratch buffer was used, dropping frame. */
+		dev_dbg(cru->dev, "Dropping frame %u\n", cru->sequence);
+	}
+
+	cru->sequence++;
+
+	/* Prepare for next frame */
+	rzg2l_cru_fill_hw_slot(cru, slot);
+
+done:
+	spin_unlock_irqrestore(&cru->qlock, flags);
+
+	return IRQ_RETVAL(handled);
+}
+
 static int rzg2l_cru_start_streaming_vq(struct vb2_queue *vq, unsigned int count)
 {
 	struct rzg2l_cru_dev *cru = vb2_get_drv_priv(vq);
@@ -637,8 +848,10 @@ static int rzg2l_cru_start_streaming_vq(struct vb2_queue *vq, unsigned int count
 		goto assert_aresetn;
 	}
 
-	ret = request_irq(cru->image_conv_irq, rzg2l_cru_irq,
-			  IRQF_SHARED, KBUILD_MODNAME, cru);
+	ret = request_irq(cru->image_conv_irq,
+			cru->info->cru_type == RZG2L_CRU_TYPE ?
+			rzg2l_cru_irq : rzv2h_cru_irq,
+			IRQF_SHARED, KBUILD_MODNAME, cru);
 	if (ret) {
 		dev_err(cru->dev, "failed to request irq\n");
 		goto assert_presetn;
@@ -646,7 +859,7 @@ static int rzg2l_cru_start_streaming_vq(struct vb2_queue *vq, unsigned int count
 
 	/* Allocate scratch buffer. */
 	cru->scratch = dma_alloc_coherent(cru->dev, cru->format.sizeimage,
-					  &cru->scratch_phys, GFP_KERNEL);
+					&cru->scratch_phys, GFP_KERNEL);
 	if (!cru->scratch) {
 		return_unused_buffers(cru, VB2_BUF_STATE_QUEUED);
 		dev_err(cru->dev, "Failed to allocate scratch buffer\n");
@@ -663,13 +876,17 @@ static int rzg2l_cru_start_streaming_vq(struct vb2_queue *vq, unsigned int count
 	}
 
 	cru->state = RZG2L_CRU_DMA_STARTING;
+
+	/* Initialize value of previous memory bank slot before streaming */
+	prev_slot[cru->id] = -1;
+
 	dev_dbg(cru->dev, "Starting to capture\n");
 	return 0;
 
 out:
 	if (ret)
 		dma_free_coherent(cru->dev, cru->format.sizeimage, cru->scratch,
-				  cru->scratch_phys);
+				cru->scratch_phys);
 free_image_conv_irq:
 	free_irq(cru->image_conv_irq, cru);
 
@@ -696,7 +913,7 @@ static void rzg2l_cru_stop_streaming_vq(struct vb2_queue *vq)
 
 	/* Free scratch buffer */
 	dma_free_coherent(cru->dev, cru->format.sizeimage,
-			  cru->scratch, cru->scratch_phys);
+			cru->scratch, cru->scratch_phys);
 
 	free_irq(cru->image_conv_irq, cru);
 	return_unused_buffers(cru, VB2_BUF_STATE_ERROR);
@@ -772,8 +989,8 @@ int rzg2l_cru_dma_register(struct rzg2l_cru_dev *cru)
 }
 
 /* -----------------------------------------------------------------------------
- * V4L2 stuff
- */
+* V4L2 stuff
+*/
 
 static const struct v4l2_format_info rzg2l_cru_formats[] = {
 	{
@@ -811,8 +1028,10 @@ static u32 rzg2l_cru_format_sizeimage(struct v4l2_pix_format *pix)
 }
 
 static void rzg2l_cru_format_align(struct rzg2l_cru_dev *cru,
-				   struct v4l2_pix_format *pix)
+				struct v4l2_pix_format *pix)
 {
+	const struct rzg2l_cru_info *info = cru->info;
+
 	if (!rzg2l_cru_format_from_pixel(pix->pixelformat))
 		pix->pixelformat = RZG2L_CRU_DEFAULT_FORMAT;
 
@@ -830,8 +1049,8 @@ static void rzg2l_cru_format_align(struct rzg2l_cru_dev *cru,
 	}
 
 	/* Limit to CRU capabilities */
-	v4l_bound_align_image(&pix->width, 320, RZG2L_CRU_MAX_INPUT_WIDTH, 1,
-			      &pix->height, 240, RZG2L_CRU_MAX_INPUT_HEIGHT, 2, 0);
+	v4l_bound_align_image(&pix->width, 320, info->max_width, 1,
+				&pix->height, 240, info->max_height, 2, 0);
 
 	pix->bytesperline = rzg2l_cru_format_bytesperline(pix);
 	pix->sizeimage = rzg2l_cru_format_sizeimage(pix);
@@ -841,25 +1060,25 @@ static void rzg2l_cru_format_align(struct rzg2l_cru_dev *cru,
 }
 
 static void rzg2l_cru_try_format(struct rzg2l_cru_dev *cru,
-				 struct v4l2_pix_format *pix)
+				struct v4l2_pix_format *pix)
 {
 	/*
-	 * The V4L2 specification clearly documents the colorspace fields
-	 * as being set by drivers for capture devices. Using the values
-	 * supplied by userspace thus wouldn't comply with the API. Until
-	 * the API is updated force fixed values.
-	 */
+	* The V4L2 specification clearly documents the colorspace fields
+	* as being set by drivers for capture devices. Using the values
+	* supplied by userspace thus wouldn't comply with the API. Until
+	* the API is updated force fixed values.
+	*/
 	pix->colorspace = RZG2L_CRU_DEFAULT_COLORSPACE;
 	pix->xfer_func = V4L2_MAP_XFER_FUNC_DEFAULT(pix->colorspace);
 	pix->ycbcr_enc = V4L2_MAP_YCBCR_ENC_DEFAULT(pix->colorspace);
 	pix->quantization = V4L2_MAP_QUANTIZATION_DEFAULT(true, pix->colorspace,
-							  pix->ycbcr_enc);
+							pix->ycbcr_enc);
 
 	rzg2l_cru_format_align(cru, pix);
 }
 
 static int rzg2l_cru_querycap(struct file *file, void *priv,
-			      struct v4l2_capability *cap)
+				struct v4l2_capability *cap)
 {
 	strscpy(cap->driver, KBUILD_MODNAME, sizeof(cap->driver));
 	strscpy(cap->card, "RZG2L_CRU", sizeof(cap->card));
@@ -868,7 +1087,7 @@ static int rzg2l_cru_querycap(struct file *file, void *priv,
 }
 
 static int rzg2l_cru_try_fmt_vid_cap(struct file *file, void *priv,
-				     struct v4l2_format *f)
+					struct v4l2_format *f)
 {
 	struct rzg2l_cru_dev *cru = video_drvdata(file);
 
@@ -878,7 +1097,7 @@ static int rzg2l_cru_try_fmt_vid_cap(struct file *file, void *priv,
 }
 
 static int rzg2l_cru_s_fmt_vid_cap(struct file *file, void *priv,
-				   struct v4l2_format *f)
+				struct v4l2_format *f)
 {
 	struct rzg2l_cru_dev *cru = video_drvdata(file);
 
@@ -893,7 +1112,7 @@ static int rzg2l_cru_s_fmt_vid_cap(struct file *file, void *priv,
 }
 
 static int rzg2l_cru_g_fmt_vid_cap(struct file *file, void *priv,
-				   struct v4l2_format *f)
+				struct v4l2_format *f)
 {
 	struct rzg2l_cru_dev *cru = video_drvdata(file);
 
@@ -903,7 +1122,7 @@ static int rzg2l_cru_g_fmt_vid_cap(struct file *file, void *priv,
 }
 
 static int rzg2l_cru_enum_fmt_vid_cap(struct file *file, void *priv,
-				      struct v4l2_fmtdesc *f)
+					struct v4l2_fmtdesc *f)
 {
 	if (f->index >= ARRAY_SIZE(rzg2l_cru_formats))
 		return -EINVAL;
@@ -932,8 +1151,8 @@ static const struct v4l2_ioctl_ops rzg2l_cru_ioctl_ops = {
 };
 
 /* -----------------------------------------------------------------------------
- * Media controller file operations
- */
+* Media controller file operations
+*/
 
 static int rzg2l_cru_open(struct file *file)
 {
diff --git a/drivers/pinctrl/renesas/Kconfig b/drivers/pinctrl/renesas/Kconfig
index 14bd55d64..6d19e4182 100644
--- a/drivers/pinctrl/renesas/Kconfig
+++ b/drivers/pinctrl/renesas/Kconfig
@@ -41,6 +41,7 @@ config PINCTRL_RENESAS
 	select PINCTRL_PFC_R8A779H0 if ARCH_R8A779H0
 	select PINCTRL_RZG2L if ARCH_RZG2L
 	select PINCTRL_RZV2M if ARCH_R9A09G011
+	select PINCTRL_RZG2L if ARCH_RZV2H
 	select PINCTRL_PFC_SH7203 if CPU_SUBTYPE_SH7203
 	select PINCTRL_PFC_SH7264 if CPU_SUBTYPE_SH7264
 	select PINCTRL_PFC_SH7269 if CPU_SUBTYPE_SH7269
-- 
2.25.1

